pkg/eval/compile_effect.go

```diff
@@ -497,21 +497,31 @@ type invalidFD struct{ fd int }
 
 func (err invalidFD) Error() string { return fmt.Sprintf("invalid fd: %d", err.fd) }
 
+func (op *redirOp) fileChan() chan interface{} {
+	if op.mode == parse.Read {
+		// ClosedChan is used when redirecting from a named file because we
+		// don't support reading "values" (as opposed to bytes) from a file.
+		return ClosedChan
+	}
+	// TODO: Replace BlackholeChan for output with a value sink that will
+	// throw an exception if it sees any values since we don't support writing
+	// values to a file without first converting them to bytes.
+	return BlackholeChan
+}
+
 func (op *redirOp) exec(fm *Frame) error {
 	var dst int
-	if op.dstOp == nil {
-		// use default dst fd
+	if op.dstOp == nil { // the common case: use default destination file-descriptor
 		switch op.mode {
 		case parse.Read:
-			dst = 0
+			dst = 0 // stdin
```

Comment:
I don't find it really useful to document that 0 is stdin and 1 is stdout; their use is too common in the codebase for such comments to be useful.

Reply:
The problem here is that `dst` is ambiguous. When I first read this function the magic numbers, in light of the ambiguous var name, was briefly confusing. Something like `dstFd` would be clearer. Too, while anyone who has programmed on UNIX for a non-trivial amount of time recognizes that 0==stdin and 1==stdout (and 2==stderr) in a context that takes a file descriptor, I'm not a fan of using those magic literals. You probably wouldn't use `10` as a fd literal even though in the context of your program it has a well defined meaning. You would probably give it a symbolic name.